### PR TITLE
RDM-13033: Fix Transactional issues

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/AuthorisedCreateEventOperation.java [hunk 1]
```diff
@@ -80,6 +86,14 @@ public CaseDetails createCaseEvent(String caseReference,
         return verifyReadAccess(caseTypeDefinition, accessProfiles, caseDetails);
     }
 
+    private void verifyCaseAccessCategories(Set<AccessProfile> accessProfiles, CaseDetails existingCaseDetails) {
+        Optional<CaseDetails> filteredCaseDetails = Optional.of(existingCaseDetails)
+            .filter(caseAccessCategoriesService.caseHasMatchingCaseAccessCategories(accessProfiles, false));
+        if (filteredCaseDetails.isEmpty()) {
+            throw new ResourceNotFoundException(NO_CASE_TYPE_FOUND);
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/AuthorisedCreateEventOperation.java:93]: Maybe for this scenario we should throw something else as it's more that the user doesn't have access to the case - may make more sense to stay consistent with elsewhere e.g. [here](https://github.com/hmcts/ccd-data-store-api/blob/63b846d297153af1a9fa22b3251b15c9654d52fc/src/main/java/uk/gov/hmcts/ccd/v2/external/controller/CaseController.java#L115)
```suggestion
            throw new CaseNotFoundException(caseDetails.getReferenceAsString());
```

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/CaseAccessCategoriesServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    void shouldReturnTrueWhenCreateCase() {
+        AccessProfile accessProfile = mock(AccessProfile.class);
+        Predicate<CaseDetails> caseDetailsPredicate = caseAccessCategoriesService
+            .caseHasMatchingCaseAccessCategories(Sets.newHashSet(accessProfile), true);
+        CaseDetails caseDetails = new CaseDetails();
+        assertTrue(caseDetailsPredicate.test(caseDetails));
+    }
+
+    @Test
+    void shouldReturnTrueWhenAccessProfilesAreEmpty() {
```
Comment (danlysiak) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/CaseAccessCategoriesServiceTest.java:39]: If there are no access profiles, I think we should return `false`? Presumably access control elsewhere would always prevent access anyway, but better to be safe!

File: src/test/java/uk/gov/hmcts/ccd/domain/service/createevent/AuthorisedCreateEventOperationTest.java [hunk 3]
```diff
@@ -344,4 +361,112 @@ void shouldFailIfNoCaseReadAccess() {
...
+        doReturn(classifiedCase).when(createEventOperation).createCaseEvent(CASE_REFERENCE,
+            newCaseDataContent);
+
+        assertThrows(ResourceNotFoundException.class, () ->
+            authorisedCreateEventOperation.createCaseEvent(CASE_REFERENCE, newCaseDataContent));
+    }
+
+    @Test
+    @DisplayName("should return true if case data match with case access categories")
+    void shouldReturnTrueIfCaseAccessCategoriesMatchCaseData() {
```
Comment (danlysiak) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/createevent/AuthorisedCreateEventOperationTest.java:388]:
```suggestion
    @DisplayName("should return case details if case data match with case access categories")
    void shouldReturnCaseDetailsIfCaseAccessCategoriesMatchCaseData() {
```

**Review by danlysiak — COMMENTED** [review]
Couple of minor comments, looks good overall though 👍

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/ClassifiedCreateEventOperation.java [hunk 4]
```diff
@@ -25,14 +23,13 @@ public ClassifiedCreateEventOperation(@Qualifier("default") CreateEventOperation
...
     @Override
     public CaseDetails createCaseEvent(String caseReference,
                                        CaseDataContent content) {
         final CaseDetails caseDetails = createEventOperation.createCaseEvent(caseReference,
                                                                            content);
         return Optional.ofNullable(caseDetails)
                        .flatMap(classificationService::applyClassification)
-                        .orElseThrow(() -> new CaseNotFoundException(caseReference));
+                       .orElse(null);
     }
```
Comment (Thor-tech-of-metal) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/ClassifiedCreateEventOperation.java:34]: Can we avoid returning null ?  It seems that a comuser of  createCaseEvent() would need to use the result to make a decision . would it be ok returning an  Optional .  Not mandatory .  Suggestion

**Review by Thor-tech-of-metal — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/ClassifiedCreateEventOperation.java (hunk 4 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/ClassifiedCreateEventOperation.java:34]: This was the existing logic. better to keep this logic for now

**Pull request closed without merging**